### PR TITLE
chore: Add Nx interactive executor

### LINKED ISSUE #254: Issue with "nx prisma db migrate reset"
**Describe the bug**
In document `db.md`, line 18, it instructs the user to run command `nx primsa db reset`, 
However, when itruns, it waits for confirmation from the user. Unfortunately, the confirmation prompt is not visible. So it appears that the routine never finishes. It does not become visible until the user presses ctrl-c.  

**To Reproduce**

1. In a terminal run `nx prisma db migrate reset`

**Expected behavior**
Expect prompt to appear, or alternatively specify an argument to avoid the confirmation prompt.

**Environment (please complete the following information):**
- Device: Laptop
- OS: Windows 11 Home Edition

**Recommended Solution**
The following change has been tested and resolves the problem.
1. To scaffold the database schema and seed the database with data, in a terminal run:

    ```bash
    nx prisma db migrate reset --force
    ```

### PR DESCRIPTION
**Pull request by ilearner777**
## What has changed

Updated nx-plugin executor

## Connected Issues

closes #254 
## QA Steps

1. run `nx prisma db migrate reset`
2. Ensure the following prompt is visible: `? Are you sure you want to reset your database? All data will be lost. » (y/N)`

## Post-Deployment

- [x ] Nothing required

### CODE REVIEW

**Comment by Pertempto** [comment]
@ilearner777 It's better to list your issues/questions in the comments, instead of the description. That way they will stand out by themselves, and are much easier to find.

I'm going to defer to @arrocke, since he has much more `nx` experience.

**Comment by ilearner777** [comment]
> @ilearner777 It's better to list your issues/questions in the comments, instead of the description. That way they will stand out by themselves, and are much easier to find.
> 
> I'm going to defer to @arrocke, since he has much more `nx` experience.

Thanks. Will do.

**Comment by ilearner777** [comment]
### Comments

 1. Copied code from [custom nx executor](https://gist.github.com/MinskLeo/749311c117996cf88e5594e61a4bb333).
 2. Made minor modifications for debugging and to harmonize with our naming conventions
 3. Added a sync version of `imp.ts` called `implsync.ts `which is not being used, just in case default `imp.ts` did not work. However the default does work, so it can be deleted from PR.
 4. Had to temporarily install package: `dotenv-cli` in order to read in the `.env.local` file. If we need this it can be added as a dependency.

**Comment by ilearner777** [comment]
### Issue

The prompt: `√ Are you sure you want to reset your database? All data will be lost. ...` appears and responds to user input. The command completes successfully, however, command: `nx run db:seed` does not get executed after the reset is complete.

I could change the `run-command` to `run-commands` and enable it to run multiple commands employing an array of strings rather than a single string. 

However, as I am still learning Nx, I am not sure the best way to implement the new run-command in the Nx environment.
Your guidance is appreciated.

**Comment by arrocke** [comment]
Nice job figuring out a solution! With a couple tweaks, this should work great!

I'm pretty sure the seed script is not running because the `prisma` is not being executed from the root directory. prisma looks for a seed script in the package.json in the current directory, which for us is in the root, but the prisma command now has a different working directory.

The other problem here is that we need to be able to pass arguments to the prisma command which is what our custom nx executor in the packages/nx-plugin package is supposed to do. I suggest we modify the executor in the nx-plugin package with the changes that you've got working in the new executor. Then we can remove that executor since we no longer need it.

**Comment by ilearner777** [comment]
I have a dilemma I do not know how to resolve.

### Steps Peformed

In `packages/db/project.json` I have added a new target called `prisma-reset` so I can run both the orginal `nx-plugin:prisma` and the new `tools/executors/workspace:run-command` as follows:

```json
    "prisma": {
      "executor": "@translation/nx-plugin:prisma",
      "options": {
        "schema": "packages/db/src/schema.prisma"
      }
    },
    "prisma-reset": {
      "executor": "./tools/executors/workspace:run-command",
      "options": {
        "command": "npx dotenv-cli -e packages/db/.env.local -- prisma migrate reset",
        "schema": "packages/db/src/schema.prisma",
        "cwd": "."
      }
    }
  },
```

The problem I am encountering is that this new code is designed such that it will only run static commands defined in the `options:command` line.  It will not pick up additional command line arguments.  Therefore it cannot be used to run dynamic commands like the original `db:prisma`.

So to overcome this problem, I tried to modify the `nx-plugin` file `executor.ts` file as you suggested, by replacing `execSync()` with `await execa.command()` as shown below:

```typescript
import { PrismaExecutorSchema } from './schema';
import { execa } from 'execa';

export default async function runExecutor({
  _,
  schema,
  ...options
}: PrismaExecutorSchema) {
  // The options are for executors are parsed with yargs, so arguments without a `-` or `--` are added to the `_` property.
  // We assume these are all prisma commands like `prisma migrate reset`.
  // Then we add the schema argument and any remaining standard arguments.
  const mycommand = `npx prisma ${_.join(
    ' '
  )} --schema=${schema} ${Object.entries(options)
    .map(([key, value]) => `--${key}=${value}`)
    .join(' ')}`;

  // `inherit` allows the parent terminal to send and receive input and output on stdin/out/err.
  // This is necessary because prisma commands are interactive.
  // execSync(command, {
  //   stdio: 'inherit',
  // });

  console.info(`Executing mycommand: ${mycommand}`);
  console.info(`Current working directory: ${process.cwd()}`);

  await execa.command(mycommand, {
    cwd: process.cwd(),
    stdio: [process.stdin, process.stdout, 'pipe'],
  });

  return {
    success: true,
  };
}
```

When I ran this, I got an error that appears to be telling me that some variable `command` is undefined.  But I don't know which variable `command` it is complaining about. I purposely changed the original variable `command` to `mycommand` to avoid any confusion. I was unable to find any online help to resolve this error.  See terminal output:

```bash
$ nx prisma db migrate reset --verbose

> nx run db:prisma migrate reset

Executing mycommand: npx prisma migrate reset --schema=packages/db/src/schema.prisma 
Current working directory: C:\Users\ilear\Documents\GitHub\gloss-translation
 >  NX   Cannot read properties of undefined (reading 'command')
TypeError: Cannot read properties of undefined (reading 'command')
    at runExecutor (C:\Users\ilear\Documents\GitHub\gloss-translation\packages\nx-plugin\src\executors\prisma\executor.ts:21:24)
    at C:\Users\ilear\Documents\GitHub\gloss-translation\node_modules\nx\src\command-line\run\run.js:81:23
    at Generator.next (<anonymous>)
    at fulfilled (C:\Users\ilear\Documents\GitHub\gloss-translation\node_modules\tslib\tslib.js:114:62)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)

 ——————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————— 

 >  NX   Running target prisma for project db failed
```
Interestingly, I noticed that the original executor was `execSync()`, so for fun I tried the synchronous version `implsync.ts` modifying `executor.ts` as follows:

```typescript
import { PrismaExecutorSchema } from './schema';
import { commandSync } from 'execa';

export default async function runExecutor({
  _,
  schema,
  ...options
}: PrismaExecutorSchema) {
  // The options are for executors are parsed with yargs, so arguments without a `-` or `--` are added to the `_` property.
  // We assume these are all prisma commands like `prisma migrate reset`.
  // Then we add the schema argument and any remaining standard arguments.
  const command = `npx prisma ${_.join(
    ' '
  )} --schema=${schema} ${Object.entries(options)
    .map(([key, value]) => `--${key}=${value}`)
    .join(' ')}`;

  console.info(`Executing nx:prisma command: ${command}`);
  console.info(`Current working directory: ${process.cwd()}`);

  commandSync(command, {
    cwd: process.cwd(),
    stdio: [process.stdin, process.stdout, 'pipe'],
  });

  return {
    success: true,
  };
}
```

This worked, but like the original, it not display the interactive question.  It behaved exactly like the original `nx-plugin` version.

### Conclusion
1. I do not know how to modify original `nx-plugin:prisma` executor in a way that would resolve the interactive display issue.
2. As a work around, we could add an extra target `prisma-reset` as a static command that will display the interactive question. The terminal command would be: `nx prisma-reset db`.  If you are agreeable to this, we could change our db documentation accordingly.  In addition, if there are other interactive commands that you run on a regular basis, you could add extra targets.  It is not an ideal work-around, but at this point, I don't have enough expertise to take this any further.

1. Do you have any ideas on how to resolve the `undefined (reading 'command') error?
2. What are your recommendations for next steps?

I await your response.

**Comment by ilearner777** [comment]
Here are some screenshots from JavaScript debugger.  Not sure if this is of any help:

Screenshot_1


Screenshot_2

**Comment by arrocke** [comment]
I think the issue is that in v5 of execa, you have to do `import { command } from 'execa'`. Importing `execa` like you were (`import { execa } from 'execa'`) is only supported in newer versions of execa. If you look at the readme on v5, it suggests using `const execa = require('execa')` instead.

**Comment by ilearner777** [comment]
I updated code you suggested above as follows:

```typescript
import { PrismaExecutorSchema } from './schema';
import { command } from 'execa';

export default async function runExecutor({
.......
  console.info(`Executing mycommand: ${mycommand}`);
  console.info(`Current working directory: ${process.cwd()}`);

  await command(mycommand, {
    cwd: process.cwd(),
    stdio: [process.stdin, process.stdout, 'pipe'],
  });

```

It did run, however, the interactive prompt did not appear.

*Here is screenshot waiting for input without prompt:*

Screenshot_3

Mystery is still unresolved.

**Comment by ilearner777** [comment]
Running javascript debugger did show the difference in the process stack between the 2 nx commands:

for `nx prisma db migrate reset`

Screenshot_6

for `nx prisma-reset db`

Screenshot_7

For `nx prisma-reset db`, npx `run-executor.js` is running but calling `npx-cli.js`->`cli.js`->`index.js`, whereas, for `nx prisma db migrate reset`, npx `run-executor.js` is paused.
I don't have the expertise to understand all of this or if there is a way to resolve.

Again, await you recommendations for next steps.

**Comment by arrocke** [comment]
I tried this code and it appears to work. The issue is that while waiting for stdin, the standard ways of piping stdout in the child process are blocked. I'm not entirely sure, but using the 'data' event on on the stdout stream prevents it from being blocked.

```typescript
  // The options are for executors are parsed with yargs, so arguments without a `-` or `--` are added to the `_` property.
  // We assume these are all prisma commands like `prisma migrate reset`.
  // Then we add the schema argument and any remaining standard arguments.

  // `inherit` allows the parent terminal to send and receive input and output on stdin/out/err.
  // This is necessary because prisma commands are interactive.
  const child = spawn(
    'npx',
    [
      'prisma',
      ..._,
      `--schema=${schema}`,
      ...Object.entries(options).map(([key, value]) => `--${key}=${value}`),
    ],
    {
      stdio: ['inherit', 'pipe', 'inherit'],
    }
  );

  // But we manually console.log stdout so that it flushes when waiting for stdin.
  child.stdout.on('data', (chunk) => console.log(chunk.toString()));

  await new Promise<void>((resolve, reject) => {
    child.on('exit', (code) => {
      if (code === 0) {
        resolve();
      } else {
        reject();
      }
    });
  });

  return {
    success: true,
  };
```

**Comment by ilearner777** [comment]
### Issue

I am getting a runtime error.

I am running: Windows 11 Home Edition

Here is the code:

```typescript
import { PrismaExecutorSchema } from './schema';
import { spawn } from 'child_process';

export default async function runExecutor({
  _,
  schema,
  ...options
}: PrismaExecutorSchema) {
  // The options are for executors are parsed with yargs, so arguments without a `-` or `--` are added to the `_` property.
  // We assume these are all prisma commands like `prisma migrate reset`.
  // Then we add the schema argument and any remaining standard arguments.

  // `inherit` allows the parent terminal to send and receive input and output on stdin/out/err.
  // This is necessary because prisma commands are interactive.
  const child = spawn(
    'npx',
    [
      'prisma',
      ..._,
      `--schema=${schema}`,
      ...Object.entries(options).map(([key, value]) => `--${key}=${value}`),
    ],
    {
      stdio: ['inherit', 'pipe', 'inherit'],
    }
  );

  // But we manually console.log stdout so that it flushes when waiting for stdin.
  child.stdout.on('data', (chunk) => console.log(chunk.toString()));

  await new Promise<void>((resolve, reject) => {
    child.on('exit', (code) => {
      if (code === 0) {
        resolve();
      } else {
        reject();
      }
    });
  });

  return {
    success: true,
  };
}
```

Below is the error that I am getting.  This happens when the code attempts to execute line 29: `child.stdout.on(...);`

```bash
> nx run db:prisma migrate reset

node:events:492
      throw er; // Unhandled 'error' event
      ^
Error: spawn npx ENOENT
    at Process.ChildProcess._handle.onexit (node:internal/child_process:283:19)
    at onErrorNT (node:internal/child_process:476:16)
    at processTicksAndRejections (node:internal/process/task_queues:82:21)
Emitted 'error' event on ChildProcess instance at:
    at Process.ChildProcess._handle.onexit (node:internal/child_process:289:12)
    at onErrorNT (node:internal/child_process:476:16)
    at processTicksAndRejections (node:internal/process/task_queues:82:21) {
  errno: -4058,
  code: 'ENOENT',
  syscall: 'spawn npx',
  path: 'npx',
  spawnargs: [
    'prisma',
    'migrate',
    'reset',
    '--schema=packages/db/src/schema.prisma'
  ]
}
Node.js v18.17.1
```

### Summary
From what I could find, this error means that it cannot find the path to my `npx` routine.  However I have it installed and can run it from a `Git Bash` terminal. I even tried replacing `npx` with the full path: `/c/Program Files/nodejs/npx` and `/c/Progra~1/nodejs/npx` without any success. 

So I am not sure why I am getting this error. 

Again, your help is appreciated.

**Comment by arrocke** [comment]
Hmm. There's not much I can do to help troubleshoot on a windows machine. I don't have one to test on. @Pertempto, would you be able to test this code on your machine?

**Comment by Pertempto** [comment]
Sure I can try it out. No promises on how soon I can get to it tho

**Comment by ilearner777** [comment]
@arrocke I have a hunch as to why it isn't working on my windows laptop.  I have `nvm for windows` installed.  When I did that, it in created a symbolic link from C:\Program Files\nodejs -> C:\Program Files\nvm\v18.17.1.  

So my belief is that the spawned terminal session within VS Code is unable to transverse the symbolic link.  However, I am not sure why that would be as I can access `npx` fine when I run it in a terminal session directly.

I will have to remove `nvm` and try with just `nodejs` installed directly to see if that resolves the problem.
Will give you an update when I get around to trying that.

**Comment by Pertempto** [comment]
I don't want to quit using nvm

**Comment by ilearner777** [comment]
I created a new` nodejs2` directory under `Program Files` without symbolic link.  Unfortunately that did not work either. So symbolic link is not the issue.

However, that leaves me without an explanation as to  why `npx` doesn't work whereas I have tested other commands like `netstat` in place 'npx' that worked.

So I am stumped on this one.

File: packages/nx-plugin/src/executors/prisma/executor.ts
Comment (arrocke) [inline packages/nx-plugin/src/executors/prisma/executor.ts:24]: Can you try adding `shell: true` to these options and report what happens?

Comment (Pertempto) [reply packages/nx-plugin/src/executors/prisma/executor.ts:24]: That seemed to fix it for me!

**Comment by ilearner777** [comment]
@Pertempto Can you explain what you did to get it to work?

**Comment by Pertempto** [comment]
> @Pertempto Can you explain what you did to get it to work?

I added `shell: true` on line 25

https://github.com/arrocke/gloss-translation/pull/299#discussion_r1434426910

**Comment by ilearner777** [comment]
@Pertempto Thanks!  Yes that worked for me too!

I will updated the PR.

**Review by Pertempto — APPROVED** [review]
This works well for me on Windows 11 :+1:

@arrocke can you please double check that it still works on your system before merging?

File: package-lock.json
Comment (arrocke) [inline package-lock.json:1]: We should reset this file since no dependencies were added.

**Review by arrocke — APPROVED** [review]
Works for me! If you take care of this one issue, then I'll merge

**Comment by ilearner777** [comment]
Curious why this hasn't been merged yet.

File: package-lock.json
Comment (ilearner777) [reply package-lock.json:1]: Done!

**Review by arrocke — APPROVED** [review]

**Pull request merged**
